File: Leetcode_solutions/best-time-to-buy-and-sell-stock.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
  (maxCur = 0), (maxSoFar = 0);
  for (let i = 1; i < prices.length; i++) {
    maxCur = Math.max(0, (maxCur += prices[i] - prices[i - 1]));
    maxSoFar = Math.max(maxCur, maxSoFar);
  }
  return maxSoFar;
};

console.log(maxProfit([7, 1, 5, 3, 6, 4]));
console.log(maxProfit([7, 6, 4, 3, 1]));
console.log(maxProfit([2, 4, 1]));

File: Leetcode_solutions/Best_Time_to_Buy_and_Sell_Stock.py
import math


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        minprice = math.inf
        maxprofit = 0
        for i in range(len(prices)):
            if prices[i] < minprice:
                minprice = prices[i]
            elif prices[i] - minprice > maxprofit:
                maxprofit = prices[i] - minprice
        return maxprofit


import math


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if not prices:
            return 0
        current = 0
        maxsofar = 0
        for i in range(1, len(prices)):
            current = max(current + prices[i] - prices[i - 1], 0)
            maxsofar = max(maxsofar, current)
        return maxsofar
